### PR TITLE
Add flusher functionality Batcher

### PR DESCRIPTION
**Pull request by otoolep**
This allows the Batcher to be flushed on request, so when client code is terminating it allows pending data to be processed, minimizing any data loss.

This change was more extensive than I initially envisioned because I had to add the "flush" channel as a type attribute. And since I was adding that as a type attribute, I decided to make all of them attributes.

### CODE REVIEW

**Comment by otoolep** [comment]
@corylanou

File: batcher_test.go [hunk 1]
```diff
@@ -36,21 +36,43 @@ func TestBatch_Timeout(t *testing.T) {
...
 	if len(batch) != batchSize {
 		t.Errorf("received batch has incorrect length exp %d, got %d", batchSize, len(batch))
 	}
 	checkPointBatcherStats(t, batcher, -1, batchSize, 0, 1)
 }
 
+// TestBatch_Flush ensures that a batcher generates a batch when flushed
+func TestBatch_Flush(t *testing.T) {
+	batchSize := 2
+	batcher := NewPointBatcher(batchSize+1, time.Hour)
```
Comment (corylanou) [inline batcher_test.go]: Should it be batchSize=1?  Not sure I understand the +1 as it makes the batchsize 3 and then you test for 1 but initially set your var to 2.  Cut and paste?

**Comment by corylanou** [comment]
This looks good.  Minor comment but I could be mis-reading the code.  +1

**Comment by pauldix** [comment]
Do you think it's worth it to add a test that has something pull off the batcher and sleeps for 10ms, and see what its throughput is over 100ms to verify correctness?

**Comment by otoolep** [comment]
Sure @pauldix -- I can add a test like that.

**Comment by otoolep** [comment]
@corylanou -- I'm going to try a slightly different way of doing this. New PR may be coming.

File: batcher_test.go (hunk 1 above)
Comment (otoolep) [inline batcher_test.go]: Yes, this looks like a minor typo. I will fix it up.

**Comment by otoolep** [comment]
Closing in favour of https://github.com/influxdb/influxdb/pull/2695

**Pull request merged**